File: adapters/curl-impersonate/internal/redis/redis_test.go
// SPDX-License-Identifier: Apache-2.0

package redis

import (
	"context"
	"encoding/json"
	"errors"
	"testing"
	"time"

	"github.com/go-redis/redismock/v9"
)

func sampleMetadata() SessionMetadata {
	return SessionMetadata{
		SessionID:         "session-1",
		Adapter:           AdapterName,
		AdapterInstanceID: "aaaa",
		CreatedAt:         "2026-04-28T00:00:00.000Z",
		LastActiveAt:      "2026-04-28T00:00:00.000Z",
		Metadata:          map[string]any{},
	}
}

func TestPing(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()
	mock.ExpectPing().SetVal("PONG")

	c := NewClient(rdb)
	if err := c.Ping(context.Background()); err != nil {
		t.Fatalf("Ping: %v", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestSetSession(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	meta := sampleMetadata()
	body, err := json.Marshal(meta)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	mock.ExpectSet("session:"+AdapterName+":session-1", body, time.Hour).SetVal("OK")

	c := NewClient(rdb)
	if err := c.SetSession(context.Background(), "session-1", meta); err != nil {
		t.Fatalf("SetSession: %v", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestGetSessionExisting(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	meta := sampleMetadata()
	body, err := json.Marshal(meta)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	mock.ExpectGet("session:" + AdapterName + ":session-1").SetVal(string(body))

	c := NewClient(rdb)
	got, err := c.GetSession(context.Background(), "session-1")
	if err != nil {
		t.Fatalf("GetSession: %v", err)
	}
	if got == nil {
		t.Fatal("expected non-nil metadata")
	}
	if got.AdapterInstanceID != meta.AdapterInstanceID {
		t.Fatalf("unexpected metadata: %+v", got)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestGetSessionMissingReturnsNil(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	mock.ExpectGet("session:" + AdapterName + ":ghost").RedisNil()

	c := NewClient(rdb)
	got, err := c.GetSession(context.Background(), "ghost")
	if err != nil {
		t.Fatalf("GetSession: %v", err)
	}
	if got != nil {
		t.Fatalf("expected nil metadata for missing key, got %+v", got)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestGetSessionRedisErrorPropagates(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	wanted := errors.New("network down")
	mock.ExpectGet("session:" + AdapterName + ":session-1").SetErr(wanted)

	c := NewClient(rdb)
	if _, err := c.GetSession(context.Background(), "session-1"); err == nil {
		t.Fatal("expected error, got nil")
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestDeleteSession(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	mock.ExpectDel("session:" + AdapterName + ":session-1").SetVal(1)

	c := NewClient(rdb)
	if err := c.DeleteSession(context.Background(), "session-1"); err != nil {
		t.Fatalf("DeleteSession: %v", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestDeleteSessionMissingIsNoop(t *testing.T) {
	rdb, mock := redismock.NewClientMock()
	defer rdb.Close()

	// DEL on a missing key returns 0, not an error.
	mock.ExpectDel("session:" + AdapterName + ":ghost").SetVal(0)

	c := NewClient(rdb)
	if err := c.DeleteSession(context.Background(), "ghost"); err != nil {
		t.Fatalf("DeleteSession: %v", err)
	}
	if err := mock.ExpectationsWereMet(); err != nil {
		t.Fatalf("expectations: %v", err)
	}
}

func TestFromURLInvalid(t *testing.T) {
	if _, err := FromURL("not a url"); err == nil {
		t.Fatal("expected parse error")
	}
}

File: adapters/playwright/src/redis.test.ts
// SPDX-License-Identifier: Apache-2.0

import RedisMock from "ioredis-mock";
import { describe, expect, it } from "vitest";

import {
  ADAPTER_NAME,
  RedisClient,
  SESSION_TTL_SECONDS,
  type SessionMetadata,
} from "./redis.js";

type MockRedis = InstanceType<typeof RedisMock>;

const sampleMetadata = (overrides: Partial<SessionMetadata> = {}): SessionMetadata => ({
  session_id: "session-1",
  adapter: ADAPTER_NAME,
  adapter_instance_id: "aaaa",
  created_at: "2026-04-28T00:00:00.000Z",
  last_active_at: "2026-04-28T00:00:00.000Z",
  metadata: {},
  ...overrides,
});

const newClient = (): { client: RedisClient; raw: MockRedis } => {
  const raw = new RedisMock();
  return { client: new RedisClient(raw as unknown as never), raw };
};

describe("RedisClient", () => {
  it("ping resolves when redis is reachable", async () => {
    const { client } = newClient();
    await expect(client.ping()).resolves.toBeUndefined();
  });

  it("setSession writes the JSON document under session:<adapter>:<id>", async () => {
    const { client, raw } = newClient();
    const meta = sampleMetadata();

    await client.setSession("session-1", meta);

    const stored = await raw.get(`session:${ADAPTER_NAME}:session-1`);
    expect(stored).not.toBeNull();
    expect(JSON.parse(stored ?? "{}")).toEqual(meta);
  });

  it("setSession applies the 1-hour TTL", async () => {
    const { client, raw } = newClient();
    await client.setSession("session-1", sampleMetadata());

    const ttl = await raw.ttl(`session:${ADAPTER_NAME}:session-1`);
    // ioredis-mock returns the TTL in seconds; allow a small slack
    // for any test-runner hiccup.
    expect(ttl).toBeGreaterThan(SESSION_TTL_SECONDS - 5);
    expect(ttl).toBeLessThanOrEqual(SESSION_TTL_SECONDS);
  });

  it("getSession returns the parsed document for an existing key", async () => {
    const { client } = newClient();
    const meta = sampleMetadata({ adapter_instance_id: "bbbb" });
    await client.setSession("session-1", meta);

    const fetched = await client.getSession("session-1");
    expect(fetched).toEqual(meta);
  });

  it("getSession returns null for a missing key", async () => {
    const { client } = newClient();
    expect(await client.getSession("ghost")).toBeNull();
  });

  it("deleteSession removes the key", async () => {
    const { client, raw } = newClient();
    await client.setSession("session-1", sampleMetadata());
    expect(await raw.get(`session:${ADAPTER_NAME}:session-1`)).not.toBeNull();

    await client.deleteSession("session-1");
    expect(await raw.get(`session:${ADAPTER_NAME}:session-1`)).toBeNull();
  });

  it("deleteSession is a no-op on a missing key", async () => {
    const { client } = newClient();
    await expect(client.deleteSession("ghost")).resolves.toBeUndefined();
  });
});
